modules/phase_field/src/postprocessors/FeatureFloodCount.C

```diff
@@ -268,39 +268,139 @@ void FeatureFloodCount::communicateAndMerge()
+    deserialize(recv_buffers);
+    recv_buffers.clear();
 
-  /*********************************************************************************
-   *********************************************************************************
-   * End Parallel Communication Section
-   *********************************************************************************
-   *********************************************************************************/
+    mergeSets(true);
 
-  mergeSets(true);
+    /**
+     * Perform a sort to give a parallel unique sorting to the identified features.
+     * We use the "min_entity_id" inside each feature to assign it's position in the
+     * sorted vector.
+     */
+    std::sort(_feature_sets.begin(), _feature_sets.end());
+
+#ifndef NDEBUG
+    /**
+     * Sanity check. Now that we've sorted the flattened vector of features
+     * we need to make sure that the counts vector still lines up appropriately
+     * with each feature's _var_index.
+     */
+    unsigned int feature_offset = 0;
+    for (auto map_num = decltype(_maps_size)(0); map_num < _maps_size; ++map_num)
+    {
+      // Skip empty map checks
+      if (_feature_counts_per_map[map_num] == 0)
+        continue;
+
+      // Check the begin and end of the current range
+      auto range_front = feature_offset;
+      auto range_back = feature_offset + _feature_counts_per_map[map_num] - 1;
+
+      mooseAssert(range_front <= range_back && range_back < _feature_count, "Indexing error in feature sets");
+
+      if (_feature_sets[range_front]._var_idx != map_num || _feature_sets[range_back]._var_idx != map_num)
+        mooseError("Error in _feature_sets sorting, map index: " << map_num);
+
+      feature_offset += _feature_counts_per_map[map_num];
+    }
+#endif
+  }
+}
+
+void
+FeatureFloodCount::buildLocalToGlobalIndices(std::vector<std::vector<unsigned int> > & local_to_global_all) const
+{
+  mooseAssert(processor_id() == 0, "This method must only be called on the root processor");
+
+  local_to_global_all.resize(_n_procs, std::vector<unsigned int>(_max_local_size));
+
+  for (decltype(_feature_sets.size()) i = 0, end_index = _feature_sets.size(); i < end_index; ++i)
+  {
+    // Get the local indices from the feature and build a map
+    for (const auto & local_index_pair : _feature_sets[i]._orig_ids)
+    {
+      mooseAssert(local_index_pair.first < _n_procs, local_index_pair.first << ", " << _n_procs);
+      mooseAssert(local_index_pair.second < _max_local_size, local_index_pair.second << ", " << _max_local_size);
+
+                              // rank                 // local index
+      local_to_global_all[local_index_pair.first][local_index_pair.second] = i;
+    }
+  }
+}
+
+void
+FeatureFloodCount::scatterIndices(std::vector<std::vector<unsigned int> > & local_to_global_all)
+{
+  // First send _feature_count and _max_local_size for sizing vectors
+
+  std::pair<unsigned int, unsigned int> sizes_pair(_feature_count, _max_local_size);
```

Comment:
Good question. Now that @jwpeterson has merged my PR in libMesh this will get rewritten so it'll be the minimum (correct) size for each processor. I created a version of Scatter that takes non-uniform vectors.